blazar/plugins/floatingips/floatingip_plugin.py

```diff
@@ -268,6 +266,9 @@ def on_end(self, resource_id, lease=None):
         allocations = db_api.fip_allocation_get_all_by_values(
             reservation_id=fip_reservation['reservation_id'])
 
+        self.deallocate(fip_reservation, allocations)
+
+    def deallocate(self, fip_reservation, allocations):
```

Comment:
It seems to me that deallocation is an operation that can occur either before or after the reservation becomes active. In particular, one part of deallocation is deleting the allocation entry; there is just an extra step when the reservation is additionally active.

Since the fip_reservation is passed in as an argument, would it make sense to do somethig like this?

```python
def deallocate(self, fip_reservation, allocations):
    is_active = fip_reservation['status'] == status.reservation.ACTIVE
    for fip_alloc in allocations:
        if is_active:
            # remove it from the pool
        # destroy the allocation
```

This makes it more of a general operation, which could be called at various times in the reservation lifecycle or by other code. Also is pretty easily testable. But, I'm not sure if this interacts badly w/ the 3rd party plugins, which introduces the concept of a deallocate function as well?

Reply:
This is a bit tricky, since the status is not kept in the floating IP reservation. I updated it to get the base reservation, and it also avoids creating the fip pool unless the reservation is active. This adds a bit of overhead (from N to 2N) since allocations are iterated through twice, but in most cases there is only one allocation anyway.